### PR TITLE
replication: add bootstrap_strategy "supervised"

### LINKED ISSUE #8509: box.cfg{bootstrap_strategy="supervised"}
Continuation of https://github.com/tarantool/tarantool/issues/7999.
Expand `box.cfg{bootstrap_strategy=...}` with option `supervised`

Design is described [here](https://www.notion.so/tarantool/replication_connect_quorum-rework-RCQ-a749e1e6b8934f3388b2176ab52c4300)

### PR DESCRIPTION
**Pull request by sergepetrenko**
This commit adds another possible bootstrap_strategy to accompany "auto" and "config": "supervised".

Such a strategy may be useful to pin the desired bootstrap leader on an active cluster (so that the user may join the replicas from the desired node without changing their box.cfg) or to manually set the bootstrap leader among the nodes that managed to start without issues.

More details are in the docbot request.

Closes #8509

@TarantoolBot document
Title: new bootstrap strategy - "supervised"

The `bootstrap_strategy` configuration option may now be set to "supervised".

This strategy works as follows:
When bootstrapping a new replicaset, the nodes do not choose a bootstrap leader and wait for it to be appointed by the user: the user may appoint a bootstrap leader by calling `box.ctl.make_bootstrap_leader()` on the desired node. In order to do so the user has to start an admin console before the initial `box.cfg` call. The console may then be used to issue the `box.ctl.make_bootstrap_leader()` call.

When joining a new replica with `bootstrap_strategy` = "supervised", the replica will not choose the bootstrap_leader automatically, but will instead join to the node on which `box.ctl.make_bootstrap_leader()` was issued last.

### CODE REVIEW

**Comment by sergepetrenko** [comment]
Checkpatch fails like that:
```
ERROR:UNCOMMENTED_DEFINITION: 'before_replace_schema' definition without comment
#190: FILE: src/box/alter.h:35:
+extern struct trigger before_replace_schema;

total: 1 errors, 663 lines checked
Commit a918f1682337 ("replication: add bootstrap_strategy "supervised"") has style problems, please review.
```
I've opend an issue: https://github.com/tarantool/checkpatch/issues/54

File: test/replication-luatest/bootstrap_strategy_test.lua [hunk 1]
```diff
@@ -455,3 +457,240 @@ g_config_fail.test_bad_uri_or_uuid = function()
...
+local function make_bootstrap_leader_initial(sockname)
+    local sock, err = socket.tcp_connect('unix/', sockname)
+    t.assert_equals(err, nil, 'Connection successful')
+    local greeting = sock:read(128, SOCKET_TIMEOUT)
+    t.assert_str_contains(greeting, 'Tarantool', 'Connected to console')
+    t.assert_str_contains(greeting, 'Lua console', 'Connected to console')
+    sock:write('box.ctl.make_bootstrap_leader()\n', SOCKET_TIMEOUT)
+    local response = sock:read(8, SOCKET_TIMEOUT)
+    t.assert_equals(response, '---\n...\n', 'The call succeeded')
+end
```
Comment (Gerold103) [inline test/replication-luatest/bootstrap_strategy_test.lua:475]: You should probably close the socket before returning from the function.

File: test/replication-luatest/bootstrap_strategy_test.lua [hunk 2]
```diff
@@ -455,3 +457,240 @@ g_config_fail.test_bad_uri_or_uuid = function()
...
+        bootstrap_strategy = 'supervised',
+        replication = {
+            server.build_listen_uri('server1'),
+            server.build_listen_uri('server2'),
+        },
+        replication_timeout = 0.1,
+    }
+    for i = 1, 2 do
+        local alias = 'server' .. i
+        cg[alias] = cg.replica_set:build_and_add_server {
```
Comment (Gerold103) [inline test/replication-luatest/bootstrap_strategy_test.lua:493]: No need for whitespace before `{`.

File: test/replication-luatest/bootstrap_strategy_test.lua [hunk 3]
```diff
@@ -455,3 +457,240 @@ g_config_fail.test_bad_uri_or_uuid = function()
...
+            alias = alias,
+            box_cfg = box_cfg,
+        }
+    end
+end)
+
+g_supervised.test_wait_for_bootstrap_leader = function(cg)
+    cg.replica_set:start{wait_until_ready = false}
+    fiber.sleep(0.4)
```
Comment (Gerold103) [inline test/replication-luatest/bootstrap_strategy_test.lua:661]: You might need to add a comment why 0.4.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+## feature/replication
+
+* You may now control which node new replicas choose as a bootstrap leader
+  without touching node config: set `box.cfg.bootstrap_strategy` to
+  `'supervised'` and the nodes will only bootstrap off the node on which you
+  called `box.ctl.make_bootstrap_leader()` last.
```
Comment (p7nov) [inline changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md:6]:
```suggestion
  without touching node config. To do so, set `box.cfg.bootstrap_strategy` to
  `'supervised'`, and the nodes will only bootstrap off the node on which you
  called `box.ctl.make_bootstrap_leader()` last.
```

File: changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  This works on an empty replica set bootstrap as well: start the admin console
+  before configuring the nodes, configure the nodes
```
Comment (p7nov) [inline changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md:8]:
```suggestion
  This works on an empty replica set bootstrap as well: start the admin console
  before configuring the nodes. Then configure the nodes:
```

File: changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  ```lua
+  box.cfg{
+      bootstrap_strategy = 'supervised',
+      replication = ...,
+      listen = ...,
+  }
+  ```
+  and finally issue `box.ctl.make_bootstrap_leader()` through the admin console
+  on the node you want to promote. Everyone will bootstrap off that node
```
Comment (p7nov) [inline changelogs/unreleased/gh-8509-bootstrap-strategy-supervised.md:17]:
```suggestion
  Finally, call `box.ctl.make_bootstrap_leader()` through the admin console
  on the node you want to promote. All the nodes will bootstrap off that node
```

**Review by p7nov — APPROVED** [review]
Please see my suggestions on the changelog.

**Review by grafin — CHANGES_REQUESTED** [review]
Thanks for the patch! One comment from me on the code itself, didn't have time to review the tests yet, will be back later:)

**Comment by sergepetrenko** [comment]
@p7nov, thanks for your suggestions! All applied.

File: src/box/alter.cc [hunk 7]
```diff
@@ -3871,6 +3871,75 @@ on_replace_dd_priv(struct trigger * /* trigger */, void *event)
...
+ * a timestamp and replica_id of the node which authored the change.
+ */
+static int
+before_replace_dd_schema(struct trigger * /* trigger */, void *event)
+{
+	struct txn *txn = (struct txn *)event;
+	struct txn_stmt *stmt = txn_current_stmt(txn);
+	struct tuple *old_tuple = stmt->old_tuple;
+	struct tuple *new_tuple = stmt->new_tuple;
+	const char *key = tuple_field_cstr(new_tuple ? new_tuple : old_tuple,
```
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/alter.cc:3886]: I got you, homie!
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 1af93e731..0de262b4c 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3883,7 +3883,8 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
        struct txn_stmt *stmt = txn_current_stmt(txn);
        struct tuple *old_tuple = stmt->old_tuple;
        struct tuple *new_tuple = stmt->new_tuple;
-       const char *key = tuple_field_cstr(new_tuple ? new_tuple : old_tuple,
+       const char *key = tuple_field_cstr(new_tuple != NULL ?
+                                          new_tuple : old_tuple,
                                           BOX_SCHEMA_FIELD_KEY);
        if (key == NULL)
                return -1;
```

File: src/box/replication.cc [hunk 8]
```diff
@@ -909,15 +935,19 @@ applier_on_connect_f(struct trigger *trigger, void *event)
...
+		if (!tt_uuid_is_nil(&cfg_bootstrap_leader_uuid)) {
+			return tt_uuid_compare(&applier->uuid,
+					       &cfg_bootstrap_leader_uuid) == 0;
+		}
+		if (!uri_is_nil(&cfg_bootstrap_leader_uri)) {
+			return uri_addr_is_equal(&applier->uri,
+						 &cfg_bootstrap_leader_uri);
+		}
+	} else if (bootstrap_strategy == BOOTSTRAP_STRATEGY_SUPERVISED) {
+		return tt_uuid_compare(&applier->ballot.bootstrap_leader_uuid,
```
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/replication.cc:949]: Thanks! I always forget about `is_equal()`.

```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index bf6995a4e..a23730a3a 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -938,16 +938,16 @@ applier_is_bootstrap_leader(const struct applier *applier)
        assert(!tt_uuid_is_nil(&applier->uuid));
        if (bootstrap_strategy == BOOTSTRAP_STRATEGY_CONFIG) {
                if (!tt_uuid_is_nil(&cfg_bootstrap_leader_uuid)) {
-                       return tt_uuid_compare(&applier->uuid,
-                                              &cfg_bootstrap_leader_uuid) == 0;
+                       return tt_uuid_is_equal(&applier->uuid,
+                                               &cfg_bootstrap_leader_uuid);
                }
                if (!uri_is_nil(&cfg_bootstrap_leader_uri)) {
                        return uri_addr_is_equal(&applier->uri,
                                                 &cfg_bootstrap_leader_uri);
                }
        } else if (bootstrap_strategy == BOOTSTRAP_STRATEGY_SUPERVISED) {
-               return tt_uuid_compare(&applier->ballot.bootstrap_leader_uuid,
-                                      &applier->uuid) == 0;
+               return tt_uuid_is_equal(&applier->ballot.bootstrap_leader_uuid,
+                                       &applier->uuid);
        }
        return false;
 }
```

File: test/replication-luatest/bootstrap_strategy_test.lua (hunk 1 above)
Comment (sergepetrenko) [reply test/replication-luatest/bootstrap_strategy_test.lua:466]: Ok, done, although socket's gc metamethod closes it anyway.
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index b87614f36..d870fc4bf 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -463,6 +463,7 @@ local function make_bootstrap_leader_initial(sockname)
     sock:write('box.ctl.make_bootstrap_leader()\n', SOCKET_TIMEOUT)
     local response = sock:read(8, SOCKET_TIMEOUT)
     t.assert_equals(response, '---\n...\n', 'The call succeeded')
+    sock:close()
 end
 
 g_supervised.after_each(function(cg)
```

File: test/replication-luatest/bootstrap_strategy_test.lua (hunk 2 above)
Comment (sergepetrenko) [reply test/replication-luatest/bootstrap_strategy_test.lua:484]: Fixed:
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index d870fc4bf..8f172a5e1 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -482,7 +482,7 @@ g_supervised.before_test('test_bootstrap', function(cg)
     }
     for i = 1, 2 do
         local alias = 'server' .. i
-        cg[alias] = cg.replica_set:build_and_add_server {
+        cg[alias] = cg.replica_set:build_and_add_server{
             alias = alias,
             box_cfg = cg.box_cfg,
         }
@@ -641,7 +641,7 @@ g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
     }
     for i = 1, 2 do
         local alias = 'server' .. i
-        cg[alias] = cg.replica_set:build_and_add_server {
+        cg[alias] = cg.replica_set:build_and_add_server{
             alias = alias,
             box_cfg = box_cfg,
         }
```

File: test/replication-luatest/bootstrap_strategy_test.lua (hunk 3 above)
Comment (sergepetrenko) [reply test/replication-luatest/bootstrap_strategy_test.lua]: Thanks for pointing this out!
I decided to rewrite this test to rely less on timeout:
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index b0e94ec10..53d455c38 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -630,7 +630,7 @@ end
 
 g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
     cg.replica_set = replica_set:new{}
-    local box_cfg = {
+    cg.box_cfg = {
         bootstrap_strategy = 'supervised',
         replication_timeout = 0.1,
         replication = {
@@ -643,20 +643,37 @@ g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
         local alias = 'server' .. i
         cg[alias] = cg.replica_set:build_and_add_server{
             alias = alias,
-            box_cfg = box_cfg,
+            box_cfg = cg.box_cfg,
         }
     end
 end)
 
+local function wait_master_is_seen(replica, master, rs_id)
+    local addr = server.build_listen_uri(master.alias, rs_id)
+    local logfile = fio.pathjoin(replica.workdir, replica.alias .. '.log')
+    local query = string.format('remote master .* at unix/:%s',
+                                addr:gsub('%-', '%%-'))
+    t.helpers.retrying({}, function()
+        t.assert(replica:grep_log(query, nil, {filename = logfile}),
+                 replica.alias .. ' sees ' .. addr)
+    end)
+end
+
 g_supervised.test_wait_for_bootstrap_leader = function(cg)
     cg.replica_set:start{wait_until_ready = false}
-    fiber.sleep(0.4)
+
+    wait_master_is_seen(cg.server1, cg.server1, cg.replica_set.id)
+    wait_master_is_seen(cg.server1, cg.server2, cg.replica_set.id)
+    wait_master_is_seen(cg.server2, cg.server1, cg.replica_set.id)
+    wait_master_is_seen(cg.server2, cg.server2, cg.replica_set.id)
+
+    fiber.sleep(cg.box_cfg.replication_timeout)
+
     assert_not_booted(cg.server1)
     assert_not_booted(cg.server2)
 end

```

File: src/box/alter.cc [hunk 9]
```diff
@@ -3871,6 +3871,75 @@ on_replace_dd_priv(struct trigger * /* trigger */, void *event)
...
+		uint64_t new_ts = UINT64_MAX;
+		uint32_t new_id = UINT32_MAX;
+		/*
+		 * Assume anything can be stored in old_tuple, so do not require
+		 * it to have a timestamp or replica_id. In contrary to that,
+		 * always require new tuple to have a valid timestamp and
+		 * replica_id.
+		 */
+		if (old_tuple != NULL) {
+			const char *field = tuple_field(old_tuple, 2);
```
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/alter.cc:3903]: I tried to do this, but stumbled upon a problem:
`_schema`, unlike other system spaces, holds arbitrary key - value pairs, and occasionally multiple values.

It's more or less clear, that the second field should be called `BOX_SCHEMA_FIELD_VALUE` (and @Gerold103 has already done so in his instance names patch).
As for fields 3 and 4 - they can't share a common name.
For example, for "version" tuple (['version', 2, 11, 1]), fields 3 and 4 are something like `BOX_SCHEMA_FIELD_VERSION_MINOR`, `BOX_SCHEMA_FIELD_VERSION_PATCH`.

And for my new tuple ['bootstrap_leader_uuid', 'c0488250-f2d8-4c5e-9706-7cb70c4e434e', 1681746623593755, 1] fields 3 and 4 would be `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP` and `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID`.

Long story short, the enum names should be specific to the particular _schema tuple, and this way they become too long.

On the other hand, I could make fields 3 and 4 names `BOX_SCHEMA_FIELD_3` and `BOX_SCHEMA_FIELD_4`.
But such names are no better than using plain numbers, it seems.

I'm open to your suggestions on how to shorten the enum names. It would be better to access the fields by a proper name rather than some magic number, indeed.

I think we can change the naming schema for "_schema" tuples. There will be generic `BOX_SCHEMA_FIELD_KEY` and `BOX_SCHEMA_FIELD_VALUE`. And also `BOX_VERSION_FIELD_MAJOR/MINOR/PATCH`, `BOX_BOOTSTRAP_LEADER_UUID/TIMESTAMP/ID`.

[the remark this answers is not in the document]
Comment (grafin) [reply src/box/alter.cc:3903]: enums can hold several names for the same value, maybe we can have something like:
```c
/** _schema fields. */
enum {
        BOX_SCHEMA_FIELD_KEY = 0,
        BOX_SCHEMA_FIELD_VALUE = 1,
...
        BOX_SCHEMA_FIELD_VERSION_MINOR = 3,
        BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP = 3,
        BOX_SCHEMA_FIELD_VERSION_PATCH = 4,
        BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID = 4,
...
};
```

[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/alter.cc:3903]: Yes, but what about the length?
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 0de262b4c..5ce546b1d 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3900,16 +3900,22 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                 * replica_id.
                 */
                if (old_tuple != NULL) {
-                       const char *field = tuple_field(old_tuple, 2);
+                       const char *field = tuple_field(old_tuple,
+                               BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_ts = mp_decode_uint(&field);
-                       field = tuple_field(old_tuple, 3);
+                       field = tuple_field(old_tuple,
+                               BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_id = mp_decode_uint(&field);
                }
                if (new_tuple != NULL &&
-                   (tuple_field_u64(new_tuple, 2, &new_ts) != 0 ||
-                    tuple_field_u32(new_tuple, 3, &new_id) != 0)) {
+                   (tuple_field_u64(new_tuple,
+                                    BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP,
+                                    &new_ts) != 0 ||
+                    tuple_field_u32(new_tuple,
+                                    BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID,
+                                    &new_id) != 0)) {
                        return -1;
                }
                if (new_ts < old_ts || (new_ts == old_ts && new_id < old_id)) {
```
I can't fit such long names anywhere. And tuple_field access which took a single line now takes 2 or 3.

Omitting the _SCHEMA part (like BOX_FIELD_BOOTSTRAP_LEADER_TIMESTAMP) is still too long.

File: src/box/box.cc [hunk 10]
```diff
@@ -1832,6 +1829,49 @@ box_set_bootstrap_leader(void)
...
+}
+
+int
+box_make_bootstrap_leader(void)
+{
+	if (tt_uuid_is_nil(&INSTANCE_UUID)) {
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 "box.ctl.make_bootstrap_leader()",
+			 "promoting this instance before box.cfg() is called");
+		return -1;
```
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/box.cc:1850]: Yes, I meant the user has to issue `box.ctl.make_bootstrap_leader()` during an ongoing `box.cfg{}` call.
I tried to improve the description:
<details><summary>docbot request</summary>

Title: new bootstrap strategy - "supervised"

The `bootstrap_strategy` configuration option may now be set to
"supervised".

This strategy works as follows:
When bootstrapping a new replicaset, the nodes do not choose a bootstrap
leader automatically and instead wait for it to be appointed by the
user. The configuration will fail if no bootstrap leader is appointed
during a `replication_connect_timeout`.

In order to apppoint a bootstrap leader, the user has to issue
`box.ctl.make_bootstrap_leader()` call on the desired node. This must be
done during the initial `box.cfg()` call.

Possible ways to achieve this are:

1. In interactive mode:
```lua
fiber.create(box.cfg, desired_config)
box.ctl.make_bootstrap_leader()
```
2. Via an init script:
```lua
-- Init script:
console.listen(admin_port)
box.cfg(desired_config)
-- User console:
tarantoolctl enter admin port
> box.ctl.make_bootstrap_leader()
```

When joining a new replica with `bootstrap_strategy` = "supervised" to
an existing replica set, the replica will not choose the bootstrap
leader automatically, but will instead join to the node on which
`box.ctl.make_bootstrap_leader()` was issued last. In case such a node
isn't found after a `replication_connect_timeout`, the configuration
fails.
</details>

We can also allow the user to call `box.ctl.make_bootstrap_leader()` any time, even before the initial `box.cfg()` call. It's quite simple to implement, I just didn't think of it as an option.

The problems I see with this are:
1. Unclear what to do if user first calls `box.ctl.make_bootstrap_leader()` and then issues `box.cfg{bootstrap_strategy="auto"/"legacy"/"config"}`. Neither raising an error from box.cfg() nor ignoring the make_bootstrap_leader() call look good.
2. In general strange approach to configuration: first set some internal variable via `box.ctl.make_bootstrap_leader()` and then rely on this variable during configuration. If the user knows which node to promote **before** box.cfg, he may as well use `bootstrap_strategy "config"` and simply put self into box.cfg.bootstrap_leader.

File: src/box/alter.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/alter.cc:3903]: Ok, I checked other existing long enum names starting with `BOX_SCHEMA_*`, and in comparison to them our new names are not drastically longer:
```
BOX_FUNC_FIELD_IS_DETERMINISTIC                                                  
BOX_FK_CONSTRAINT_FIELD_PARENT_COLS                                              
BOX_FUNC_INDEX_FIELD_SPACE_ID                                                    
BOX_SESSION_SETTINGS_FIELD_VALUE                                                 
BOX_SPACE_SEQUENCE_FIELD_IS_GENERATED                                            

BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_UUID                                           
BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_TIMESTAMP                                      
BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_REPLICA_ID
```

I tried to shorten the names, but failed to find any good alternative.
I think it's better to stick with existing naming schema then. Will do.

P. S. Maybe @Gerold103 has some thoughts on how to shorten the `BOX_SCHEMA_FIELD_BOOTSTRAP_LEADER_*`?

[the remark this answers is not in the document]
Comment (Gerold103) [reply src/box/alter.cc:3903]: I wouldn't care too much and just used `BOX_SCHEMA_FIELD_VALUE + 1`, `BOX_SCHEMA_FIELD_VALUE + 2`, etc. But if you anyway want new names, then either use the long name, or maybe try `BOX_SCHEMA_FIELD_BOOT_LEADER_`, without `STRAP` suffix. I can't think of any shorter alternative.

[the remark this answers is not in the document]
Comment (grafin) [reply src/box/alter.cc:3903]: @sergepetrenko I'm ok with `BOX_SCHEMA_FIELD_VALUE + <number>`

[the remark this answers is not in the document]
Comment (sergepetrenko) [reply src/box/alter.cc:3903]: Done:
```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 7c743652e..c129b1467 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3893,6 +3893,8 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                uint32_t old_id = 0;
                uint64_t new_ts = UINT64_MAX;
                uint32_t new_id = UINT32_MAX;
+               int ts_field = BOX_SCHEMA_FIELD_VALUE + 1;
+               int id_field = BOX_SCHEMA_FIELD_VALUE + 2;
                /*
                 * Assume anything can be stored in old_tuple, so do not require
                 * it to have a timestamp or replica_id. In contrary to that,
@@ -3900,16 +3902,16 @@ before_replace_dd_schema(struct trigger * /* trigger */, void *event)
                 * replica_id.
                 */
                if (old_tuple != NULL) {
-                       const char *field = tuple_field(old_tuple, 2);
+                       const char *field = tuple_field(old_tuple, ts_field);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_ts = mp_decode_uint(&field);
-                       field = tuple_field(old_tuple, 3);
+                       field = tuple_field(old_tuple, id_field);
                        if (field != NULL && mp_typeof(*field) == MP_UINT)
                                old_id = mp_decode_uint(&field);
                }
                if (new_tuple != NULL &&
-                   (tuple_field_u64(new_tuple, 2, &new_ts) != 0 ||
-                    tuple_field_u32(new_tuple, 3, &new_id) != 0)) {
+                   (tuple_field_u64(new_tuple, ts_field, &new_ts) != 0 ||
+                    tuple_field_u32(new_tuple, id_field, &new_id) != 0)) {
                        return -1;
                }
                if (new_ts < old_ts || (new_ts == old_ts && new_id < old_id)) {
```

File: src/box/box.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (Gerold103) [reply src/box/box.cc:1856]: I don't mind any option. It simply feels like the requirement to start an admin console and do the stuff manually is an overkill.

File: src/box/replication.cc [hunk 11]
```diff
@@ -896,6 +909,19 @@ applier_on_connect_f(struct trigger *trigger, void *event)
...
+			/*
+			 * We may reuse the on_connect trigger, because it fires
+			 * only once. The applier will remain paused after the
+			 * trigger run.
+			 */
+			trigger_clear(trigger);
+			trigger_create(trigger, applier_on_ballot_update_f,
+				       applier, NULL);
+			trigger_add(&applier->on_ballot_update, trigger);
+		}
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply src/box/replication.cc]: Nice, nice. Thanks for doing that! The next step would be to drop some counters from `struct replicaset` like `applier.connected`, `applier.loading`, etc. I think they are used only when some applier changes its state, which means the needed counters could be just re-calculated from scratch in the relevant triggers. But this can be done some time later.

**Review by Gerold103 — APPROVED** [review]
Thanks for the fixes!
* After rebase you might need to update certain xregion usages - they won't need size nor type cast anymore.
* I suggest to rename `box_write_bootstrap_leader` to `box_set_bootstrap_leader_record`. It would be consistent with the `..._record` suffix used in my names patch.

**Comment by sergepetrenko** [comment]
Thanks for the review!

<details><summary>Applied the following fixes:</summary>

```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index c129b1467..8331ff5ba 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -3995,11 +3995,8 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
                        dd_version_id = tarantool_version_id();
                }
        } else if (strcmp(key, "bootstrap_leader_uuid") == 0) {
-               size_t size;
-               struct tt_uuid *uuid =
-                       (struct tt_uuid *)xregion_alloc_object(&txn->region,
-                                                              typeof(*uuid),
-                                                              &size);
+               struct tt_uuid *uuid = xregion_alloc_object(&txn->region,
+                                                           typeof(*uuid));
                if (!new_tuple) {
                        *uuid = uuid_nil;
                } else if (tuple_field_uuid(new_tuple, BOX_SCHEMA_FIELD_VALUE,
diff --git a/src/box/applier.cc b/src/box/applier.cc
index c737b68f4..67aa5f363 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -544,10 +544,9 @@ applier_ballot_data_create(struct applier_ballot_data *data)
 static int
 applier_on_first_ballot_update_f(struct trigger *trigger, void *event)
 {
-       (void)event;
+       struct applier *applier = (struct applier *)event;
        struct applier_ballot_data *data =
                (struct applier_ballot_data *)trigger->data;
-       struct applier *applier = (struct applier *)event;
        if (!diag_is_empty(diag_get()))
                diag_move(diag_get(), &data->diag);
        else if (!applier->ballot.is_booted && !applier->ballot.is_ro) {
diff --git a/src/box/box.cc b/src/box/box.cc
index a8ad3f3f0..5d9d1a1f0 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1837,7 +1837,7 @@ box_set_bootstrap_leader(void)
 
 /** Persist this instance as the bootstrap leader in _schema space. */
 static int
-box_write_bootstrap_leader(void)
+box_set_bootstrap_leader_record(void)
 {
        assert(instance_id != REPLICA_ID_NIL);
        assert(!tt_uuid_is_nil(&INSTANCE_UUID));
@@ -1868,9 +1868,8 @@ box_make_bootstrap_leader(void)
        if (is_box_configured) {
                if (box_check_writable() != 0)
                        return -1;
-               assert(instance_id != REPLICA_ID_NIL);
                /* Ballot broadcast will happen in an on_commit trigger. */
-               return box_write_bootstrap_leader();
+               return box_set_bootstrap_leader_record();
        } else {
                bootstrap_leader_uuid = INSTANCE_UUID;
                box_broadcast_ballot();
@@ -4342,7 +4341,7 @@ bootstrap_master(const struct tt_uuid *replicaset_uuid)
        /* Set UUID of a new replica set */
        box_set_replicaset_uuid(replicaset_uuid);
        if (bootstrap_strategy == BOOTSTRAP_STRATEGY_SUPERVISED)
-               box_write_bootstrap_leader();
+               box_set_bootstrap_leader_record();
 
        /* Enable WAL subsystem. */
        if (wal_enable() != 0)
```
</details>

**Comment by sergepetrenko** [comment]
This should fix the integration / cartridge test failures:
```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index 67aa5f363..9e32d5bbc 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -485,6 +485,7 @@ applier_watch_ballot(struct applier *applier)
                         */
                        if (is_empty)
                                continue;
+                       diag_clear(diag_get());
                        applier_run_ballot_triggers(applier);
                }
        } catch (ClientError *e) {
@@ -494,6 +495,7 @@ applier_watch_ballot(struct applier *applier)
 try_vote:
        applier_get_ballot_from_vote(&io, &row, &ibuf,
                                     &applier->ballot);
+       diag_clear(diag_get());
        applier_run_ballot_triggers(applier);
 }
 

```

**Comment by sergepetrenko** [comment]
Another minor change to fix occasional test hangs:
```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index 7c49243a2..deb7da580 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -726,12 +726,14 @@ replicaset_update(struct applier **appliers, int count, bool keep_connec>
                         * when it is finally connected.
                         */
                        rlist_add_entry(&anon_replicas, replica, in_anon);
+                       replica->applier_sync_state = APPLIER_DISCONNECTED;
                        continue;
                }
 
                assert(!tt_uuid_is_nil(&applier->uuid));
                replica->uuid = applier->uuid;
                replica->anon = applier->ballot.is_anon;
+               replica->applier_sync_state = APPLIER_CONNECTED;
 
                if (replica_hash_search(&uniq, replica) != NULL) {
                        replica_clear_applier(replica);

```
